### ISSUE TITLE
VNC and RDP protocol? 

### ISSUE THREAD

**Issue by Rudde** [yrutschle/sslh#95]
Would it be possible to implement VNC and Windows RDP protocol (for forwarding to a different node in the NAT)?

**Comment by yrutschle** [comment]
On Sat, Jun 18, 2016 at 06:03:33AM -0700, Rudde wrote:
> Would it be possible to implement VNC and Windows RDP protocol (for forwarding to a different node in the NAT)?

Possibly, if the client sends the first application-level
message, and that message contains data that enables to
identify the protocols.

**Comment by shujaaa** [comment]
Hi,
Just to go on on the topic.

When trying to redirect VNC, I thought it would fall under "anyprot" but apparently it is classified as SSH.

Is it normal behavior ? Have you had time to check what the protocol looks like ?

Thanks,

**Comment by Rudde** [comment]
Hey, I do not know too much about it, and didn't really know if it was possible or not. I did a packet capture from the server side of a Win RDP see attached file, and remove the .txt after download open in wireshark. 

I can do one with VNC too but is there any difference between TinyVNC RealVNC and all of those? Will it matter if I do it on Windows or Linux?


[WinRDPServer.pcapng.txt](https://github.com/yrutschle/sslh/files/347649/WinRDPServer.pcapng.txt)

**Comment by whughes339** [comment]
Hi Rudde,
I'm trying to do the same thing as you: create a regex probe for the RDP protocol, since using anyprot is a bit clunky. I sniffed out an RDP packet, but didn't see if there was anything useful in the initial packet that SSLH sees.

I wonder if there's a way to debug SSLH to print the plaintext packet to stdout. That would help to see what the RDP packet looks like from SSLH. I tried messing around with the source code, but couldn't figure anything out.

**Comment by rdebath** [comment]
RDP looks easy, I get two forms of the "client hello" ...
```
  03 00 00 ** ** e0 00 00 00 00 00 "Cookie: mstshash="
```
and
```
  03 00 00 13 0e e0 00 00 00 00 00 01 00 08 00 0b 00 00 00
```
The first one is used if `mstsc` has a user name to give to the other end, the second when it doesn't. The first part of the user name; eg: "Administr" appears after the `=`. The two `** **` values look like a packet length and some sort of offset.

Unfortunately, it seems that VNC does not have a client hello, it waits for the server to say 'RFB ' before it sends anything.

**Comment by orbitarm** [comment]
Yup, I actually found the relevant info within the official RDP specs (https://msdn.microsoft.com/en-us/library/cc240445.aspx):

> 4.1.1 Client X.224 Connection Request PDU
> The following is an annotated dump of the X.224 Connection Request PDU (section 2.2.1.1).
> ```
> 00000000 03 00 00 2c 27 e0 00 00 00 00 00 43 6f 6f 6b 69 ...,'......Cooki
> 00000010 65 3a 20 6d 73 74 73 68 61 73 68 3d 65 6c 74 6f e: mstshash=elto
> 00000020 6e 73 0d 0a 01 00 08 00 00 00 00 00 ns..........
> 03 -> TPKT Header: version = 3
> 00 -> TPKT Header: Reserved = 0
> 00 -> TPKT Header: Packet length - high part
> 2c -> TPKT Header: Packet length - low part (total = 44 bytes)
> 27 -> X.224: Length indicator (39 bytes)
> e0 -> X.224: Type (high nibble) = 0xe = CR TPDU; credit (low nibble) = 0
> 00 00 -> X.224: Destination reference = 0
> 00 00 -> X.224: Source reference = 0
> 00 -> X.224: Class and options = 0
> 43 6f 6f 6b 69 65 3a 20 6d 73 74 73 68 61 73 68
> 3d 65 6c 74 6f 6e 73 -> "Cookie: mstshash=eltons"
> 0d0a -> Cookie terminator sequence
> 01 -> RDP_NEG_REQ::type (TYPE_RDP_NEG_REQ)
> 00 -> RDP_NEG_REQ::flags (0)
> 08 00 -> RDP_NEG_REQ::length (8 bytes)
> 00 00 00 00 -> RDP_NEG_REQ: Requested protocols (PROTOCOL_RDP)
> ```

The regex is a little sketchy in sslh (I especially had trouble with recognizing \x00 bits) but this has been working as an RDP probe for me:

`{ name: "regex"; host: "yourhosthere.lan"; port: "3389"; regex_patterns: [ "^\x03[^\x01-\xFF][^\x01-\xFF]..\xE0" ]; }`

**Comment by rdebath** [comment]
It seems that doesn't always work; if you add `negotiate security layer:i:0` to the RDP file you get a TLS client hello packet instead. Complete with SNI extension.

... hummmm ... yup, `openssl s_client ...` connects successfully to the RDP port on a windows machine.

**Comment by CE1CECL** [comment]
RDP works with the SOCKS argument fine for me, FYI.

**Comment by liujunhui2** [comment]
> RDP works with the SOCKS argument fine for me, FYI.

Could you elaborate please? Where to put the SOCKS argument? Thanks.

**Comment by CE1CECL** [comment]
> > RDP works with the SOCKS argument fine for me, FYI.
> 
> Could you elaborate please? Where to put the SOCKS argument? Thanks.

Actually now there is a `--msrdp=localhost:3389` argument, back in 2023 it wasn't there. That is in the latest version anyways. IIRC I did `--socks5=localhost:3389` before. As for the order I don't remember but it can vary anyways depending on what you are hosting.
